### DISCUSSION TITLE
Linux display issues

### DISCUSSION THREAD

**Discussion by zsliu98** [ZL-Audio/ZLEqualizer#31] [General]
discuss Linux display issues here

**Comment by mikelupe** [comment]
As already mentioned, it now builds and works well, until I try to drag an EQ point - at which point it freezes the plugin and the host.

Changing the frequency with the FREQ knob doesn't create any problems though.
    Intel i7
    Debian 12 Bookworm
    KDE Plasma 5.27.5
    X11
    AMD GPU open source drivers

I'll only be able to create a stacktrace tomorrow or later.

**Comment by David-Else** [comment]
As soon as I grab an EQ point and drag it around for a while it eventually crashes and freezes the plugin. I am using Reaper, and if I start it from the CLI there is no useful error reporting. If there is anything I can do to help debug please let me know.

**Comment by mikelupe** [comment]
Btw the same issue when dragging EQ point happens on Debian 13 Trixie unstable

**Comment by zsliu98** [comment]
There are several potential reasons:
- the adjustment of freq, gain, and Q is not real-time safe, which finally leads to priority inversion, i.e., the audio thread becomes a non-real-time thread. I will refactor my code once I come up with a solution.
- the painting (especially the floating window) congests the message thread because of no hardware acceleration
- there are some unknown dead-locks in my code

I will look into it later but I am afraid none of them is easy to solve.

**Reply by mikelupe** [reply]
Thanks for your efforts - whenever it may happen :) I strongly believe this could become the numero uno open source EQ, with its appealing GUI and features.

**Comment by pms967** [comment]
Just for the record: I eventually have been able to reproduce the issue, by dragging around one point back and forth like crazy (while playing). It worked for a while, then everything got frozen. Indeed it looks like it's likely related to the problem with automation.

**Reply by swindus** [reply]
That's exactly what happens to me.

**Comment by zsliu98** [comment]
Here are some updates on this one:
- I think the automation (real-time safety) issue is the key. Some Windows users also reported freezing when they added a band or set up automation.
- Here is a fix for the automation issue (see the attachment or this [action](https://github.com/ZL-Audio/ZLEqualizer/actions/runs/8475626824)). I am not very confident about this since it is very tricky to ensure real-time safety :innocent: I will review my code again when I have time.
- I will include the LV2 format in the Linux release although it has not been tested.

@mikelupe @David-Else Would you like to try this out when you have time?

[ZL Equalizer-0.0.0.zip](https://github.com/ZL-Audio/ZLEqualizer/files/14797886/ZL.Equalizer-0.0.0.zip)

**Reply by mikelupe** [reply]
On Linux I just build latest "develop" version and pushed moving around EQ points by mouse during ~10m in Ardour 8.4.

Result: no freeze, even with highest zoom factor "full-screen" both VST3 & LV2

@David-Else can you confirm this?

EDIT: I indeed realized when playing written automation curves and having the plugin GUI displayed (LV2 & VST3), the host (Ardour) loses any display updates - so the playhead doesn't move until there's no more movement in the plugin GUI. 

But this is known and well communicated - now it doesn't freeze anymore - thanks for your efforts.

**Comment by pms967** [comment]
> @mikelupe @David-Else Would you like to try this out when you have time?
> 
> [ZL Equalizer-0.0.0.zip](https://github.com/ZL-Audio/ZLEqualizer/files/14797886/ZL.Equalizer-0.0.0.zip)

Tried both VST3 and LV2 versions, with both Reaper 7.x and Ardour 7.3 on Debian stable: works perfectly!  

No crash, even if moving around the point insanely.  Great job!

**Comment by David-Else** [comment]
I am away from my music computer for a while, but will check out the fix when I get back.

**Comment by David-Else** [comment]
I have tried [0.1.16](https://github.com/ZL-Audio/ZLEqualizer/releases/tag/0.1.16) and the crashing seems to be fixed, thanks, great work!
